### PR TITLE
add LCS algorithm (dynamic programming approach)

### PR DESCRIPTION
**Pull request by itskurtz**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: dynamic_programming/lcs.c [hunk 1]
```diff
@@ -0,0 +1,151 @@
+/**
+ * @file
+ * @brief [LCS](https://en.wikipedia.org/wiki/Longest_common_subsequence_problem) algorithm
+ * @details
+ * From Wikipedia: The longest common subsequence (LCS) problem is the problem
+ * of finding the longest subsequence common to all sequences in a set of sequences
+ * (often just two sequences).
+ * @author [Kurtz](https://github.com/itskurtz)
+ * @see dynamic_programming/lcs.c
```
Comment (Panquesito7) [inline dynamic_programming/lcs.c:9]: `@see` is only used for other similar implementations within the code.
For example, a file named `quick_sort.cpp` mentions the `quick_sort_2.cpp` implementation.

File: dynamic_programming/lcs.c [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+ */
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
```
Comment (Panquesito7) [inline dynamic_programming/lcs.c:14]: Please add a one-line description of what the library/header is for (**see the example below**).

```c
#include <stdio.h>     /// for IO operations
#include <assert.h>    /// for assert
```

File: dynamic_programming/lcs.c [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+		else
+			i = i - 1;
+	}
+	return lcs;
+}
+/**
+ * @brief test function
+ * @returns void
+ */
+void test(void) {
```
Comment (Panquesito7) [inline dynamic_programming/lcs.c:91]:
```suggestion
 * @brief Self-test implementations
 * @returns void
 */
static void test() {
```

File: dynamic_programming/lcs.c [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+	
+	printf("S1: %s\tS2: %s\n", s1, s2);
+	printf("LCS len:%3d\n", L[l1][l2]);
+	printf("LCS: %s\n", lcs);
+
+	free(lcs);	
+	for (j = 0; j <= l1; j++)
+		free(L[j]), free(B[j]);
+	free(L);
+	free(B);
```
Comment (Panquesito7) [inline dynamic_programming/lcs.c:140]:
```suggestion
	free(B);

	printf("All tests have successfully passed!\n");
```

File: dynamic_programming/lcs.c [hunk 5]
```diff
@@ -0,0 +1,151 @@
+/**
+ * @file
+ * @brief [LCS](https://en.wikipedia.org/wiki/Longest_common_subsequence_problem) algorithm
```
Comment (Panquesito7) [inline dynamic_programming/lcs.c:3]:
```suggestion
 * @brief [Longest Common Subsequence](https://en.wikipedia.org/wiki/Longest_common_subsequence_problem) algorithm
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 🚀

File: dynamic_programming/lcs.c [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+	free(lcs);	
+	for (j = 0; j <= l1; j++)
+		free(L[j]), free(B[j]);
+	free(L);
+	free(B);
+}
+
+/**
+ * @brief Main function
```
Comment (Panquesito7) [inline dynamic_programming/lcs.c:144]:
```suggestion
 * @brief Main function
 * @param argc commandline argument count (ignored)
 * @param argv commandline array of arguments (ignored)
```

File: dynamic_programming/lcs.c [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+ * @param argc commandline argument count (ignored)
+ * @param argv commandline array of arguments (ignored)
+ * @returns 0 on exit
+ */
+int main(int argc, char *argv[]) {
+	
+	test();
```
Comment (Panquesito7) [inline dynamic_programming/lcs.c:152]:
```suggestion
	test();  // run self-test implementations
```

**Review by Panquesito7 — DISMISSED** [review]
Awesome work. Thanks! 🚀

**Review by tjgurwara99 — APPROVED** [review]

**Review by Panquesito7 — APPROVED** [review]

**Pull request merged**